drivers/bluetooth/hci/userchan.c

```diff
@@ -93,6 +100,56 @@ static struct net_buf *get_rx(const uint8_t *buf)
 	return NULL;
 }
 
+/**
+ * @brief Decode the length of an HCI H4 packet
+ * @details Decodes packet length according to Bluetooth spec v5.2 Vol 4 Part E
+ * @param packet_type	Type of packet according to Bluetooth spec v5.2 Vol 4, Part A Table 2.1
+ * @param packet_buff	Pointer to packet buffer
+ */
+static uint16_t decode_hci_packet_length(uint8_t packet_type, const uint8_t *packet_buff)
+{
+	uint16_t payload_size = 0;
+	uint8_t header_size = 0;
+	const uint8_t packet_type_len = 1;
+
+	switch (packet_type) {
+	/* Bluetooth spec 5.2 Vol 4, Part E - 5.4.1 HCI Command Packet */
+	case H4_CMD:
+		/* Parameter Total Length */
+		payload_size = packet_buff[H4_CMD_PAYLOAD_SIZE_OFF];
+		header_size = BT_HCI_CMD_HDR_SIZE;
+		break;
+	/* Bluetooth spec 5.2 Vol 4, Part E - 5.4.2 ACL Data Packets */
+	case H4_ACL:
+		/* Data Total Length */
+		payload_size = sys_get_le16(&packet_buff[H4_ACL_PAYLOAD_SIZE_OFF]);
+		header_size = BT_HCI_ACL_HDR_SIZE;
+		break;
+	/* Bluetooth spec 5.2 Vol 4, Part E - 5.4.3 HCI Synchronous Data Packets */
+	case H4_SCO:
+		/* Data_Total_Length */
+		payload_size = packet_buff[H4_SCO_PAYLOAD_SIZE_OFF];
+		header_size = 3;
+	/* Bluetooth spec 5.2 Vol 4, Part E - 5.4.4 HCI Event Packet */
+	case H4_EVT:
+		/* Data_Total_Length */
+		payload_size = packet_buff[H4_EVT_PAYLOAD_SIZE_OFF];
+		header_size = BT_HCI_EVT_HDR_SIZE;
+		break;
+	/* Bluetooth spec 5.2 Vol 4, Part E - 5.4.5 HCI ISO Data Packets */
+	case H4_ISO:
+		/* ISO_Data_Load_Length parameter */
+		payload_size =  sys_get_le16(&packet_buff[H4_ISO_PAYLOAD_SIZE_OFF]) & 0x3FFF;
+		header_size = BT_HCI_ISO_HDR_SIZE;
```

Comment:
I wonder if it would make more sense to actually extract the header in each case, and use the `len` field in each header to extract the `payload_size`. That might be better than using these offsets and so on. 

So it would be something like 

```c
	case H4_ACL:
		/* Data Total Length */
		const struct bt_hci_acl_hdr *hdr = &packet_buff[sizeof(packet_type)];

		payload_size = sys_get_le16(hdr->len);
		header_size = sizeof(hdr);
		break;
```

Though it may require some endianess handling or something for this to properly work.

Reply:
Yes 👍 , I was checking the source code of how the hci structs are handled and its similar to what you mention

See for example

https://github.com/zephyrproject-rtos/zephyr/blob/2d65acca3a76b968657c2e796512558e93f2ed23/drivers/bluetooth/hci/h4.c#L127C3-L127C3.